packages/backend-app-api/src/services/implementations/auth/external/helpers.ts

```diff
@@ -0,0 +1,139 @@
+ *
+ * @internal
+ */
+export function readAccessRestrictionsFromConfig(
+  externalAccessEntryConfig: Config,
+): BackstagePrincipalAccessRestrictions | undefined {
+  const configs =
+    externalAccessEntryConfig.getOptionalConfigArray('accessRestrictions') ??
+    [];
+
+  const result: BackstagePrincipalAccessRestrictions = [];
+  for (const config of configs) {
+    const validKeys = ['plugin', 'permission', 'permissionAttribute'];
+    for (const key of config.keys()) {
+      if (!validKeys.includes(key)) {
+        const valid = validKeys.map(k => `'${k}'`).join(', ');
+        throw new Error(
+          `Invalid key '${key}' in 'accessRestrictions' config, expected one of ${valid}`,
+        );
+      }
+    }
+
+    const pluginId = config.getString('plugin');
+    const permissionNames = readPermissionNames(config);
+    const permissionAttributes = readPermissionAttributes(config);
+
+    result.push({
+      pluginId,
+      ...(permissionNames ? { permissionNames } : {}),
+      ...(permissionAttributes ? { permissionAttributes } : {}),
+    });
+  }
+
+  return result?.length ? result : undefined;
+}
+
+/**
+ * Reads a config value as a string or an array of strings, and deduplicates and
+ * splits by comma/space into a string array. Can also validate against a known
+ * set of values. Returns undefined if the key didn't exist or if the array
+ * would have ended up being empty.
+ */
+function stringOrStringArray<T extends string>(
+  root: Config,
+  key: string,
+  validValues?: readonly T[],
+): T[] | undefined {
+  if (!root.has(key)) {
+    return undefined;
+  }
+
+  const rawValues = Array.isArray(root.get(key))
+    ? root.getStringArray(key)
+    : [root.getString(key)];
+
+  const values = [
+    ...new Set(
+      rawValues
+        .map(v => v.split(/[ ,]/))
+        .flat()
+        .filter(Boolean),
+    ),
+  ];
+
+  if (!values.length) {
+    return undefined;
+  }
+
+  if (validValues?.length) {
+    for (const value of values) {
+      if (!validValues.includes(value as T)) {
+        const valid = validValues.map(k => `'${k}'`).join(', ');
+        throw new Error(
+          `Invalid value '${value}' at '${key}' in 'permissionAttributes' config, valid values are ${valid}`,
+        );
+      }
+    }
+  }
+
+  return values as T[];
+}
+
+function readPermissionNames(externalAccessEntryConfig: Config) {
+  return stringOrStringArray(externalAccessEntryConfig, 'permission');
+}
+
+function readPermissionAttributes(externalAccessEntryConfig: Config) {
+  const config = externalAccessEntryConfig.getOptionalConfig(
+    'permissionAttribute',
+  );
+  if (!config) {
+    return undefined;
+  }
+
+  const validKeys = ['action'];
+  for (const key of config.keys()) {
+    if (!validKeys.includes(key)) {
+      const valid = validKeys.map(k => `'${k}'`).join(', ');
+      throw new Error(
+        `Invalid key '${key}' in 'permissionAttribute' config, expected ${valid}`,
+      );
```

Comment:
I feel this validation settles the discussion at https://github.com/backstage/backstage/pull/24702#discussion_r1602798781 😅

Reply:
Yep